### ISSUE TITLE
Best Practice: API for models

### ISSUE THREAD

**Issue by mjreich** [nxus/core#88]
Rather than exposing a raw Waterline model for querying, we should wrap the model in an API provided by the module for accessing.

## Old style
```javascript
app.get('storage').getModel('users').((User) => {
  return User.find({email: ’something’}).populate('team')
}).then((users) => {
  …
})
```

## New style
```javascript
import {users} from 'nxus-users'

users.findUsers({email: ’something’}).populate('team').then((users) => {
   ...
})
```

Or, even better if this is a common query:

```javascript
import {users} from 'nxus-users'

users.findUsersByEmail(’something’).populate('team').then((users) => {
   ...
})
```

**Comment by loppear** [comment]
Not sure I agree on moving all Model methods up to the Module as BP - these methods typically make sense on the model (`User.findByEmail`) to me (in fact, those are already generated by waterline no?), and we want a way to have models be more directly accessible on the containing module rather than through storage.

How about
```
users.models.User.findByEmail
```
This would match the module-internal `.models` from HasModels, and in fact HasModels could create this support. In either style however, this will require some proxy work - `findUsersByEmail` or `models.User.findByEmail` both need to return a proxying promise that reconstructs the remaining chain of calls when any step along the way resolves. My approach would also require some support on ModuleProxy to allow (storage.HasModels) to add new accessors (`.models`).

**Comment by mjreich** [comment]
@loppear no, I don’t believe waterline does the nifty introspection query method trick 😦 At least not that I could find in the docs.  The thing I’d like to avoid is providing direct access to the model in case we move away from Waterline (which may be something to consider in a future version).

**Comment by loppear** [comment]
Waterline does, poorly documented indeed. My argument is that the convenience query api is something that should be provided by our choice of ORM, and is provided by our current choice of ORM.

Separate from that part of the proposal, I still agree we need a way to access a module's models-querying-methods from the moduleproxy rather than through storage.

**Comment by ScottMaxson** [comment]
Ooof - the WL docs pretend to be good (well, at least they pass the initial sniff test) but then fall down in so many places. Discouraging. ORM needs a lot of good documentation to be useful, because it's a Hard Problem.

Wrapping WL's API *could* be a quagmire, as you all know. Where do you stop? 

I do think simplifying the common app.get('storage').getModel().then() typing for each query against the object-model would be nice.

Also - think about testability too. It would be really nice to have a memory-backed test version of the ORM layer to switch over to, and this would likely be easier to expose if some of WL's details are hidden.

**Comment by loppear** [comment]
Waterline has adapters for sails-disk/sails-memory as well as for sqlite3 (which should support either disk or memory) that we should test out for running clean tests.

**Comment by mjreich** [comment]
@loppear sqlite3 is already the default for Storage and works consistently well.  I vote for that.
